google/cloud/security/common/gcp_type/firewall_rule.py

```diff
@@ -65,11 +82,166 @@ def __init__(self, **kwargs):
+
+        This rule must be valid to return the representation.
+
+        Returns:
+          str: A string JSON dump of the firewall rule.
+
+        Raises:
+          InvalidFirewallRuleError: If the firewall rule is invalid.
+          InvalidFirewallActionError: If the firewall action is invalid.
+        """
+        self.validate()
+        firewall_dict = {
+            'direction': self.direction,
+            'network': self.network,
+        }
+        for key, value in [self.firewall_action.json_key(),
+                           ('sourceRanges', list(self._source_ranges)),
+                           ('sourceTags', list(self._source_tags)),
+                           ('targetTags', list(self._target_tags)),
+                           ('destinationRanges',
+                            list(self._destination_ranges)),
+                           ('priority', self._priority)]:
+            if value:
+                firewall_dict[key] = value
+        return json.dumps(firewall_dict, sort_keys=True)
+
+    def validate(self):
+        """Validates that a rule is valid and not a duplicate.
+
+        Validation is based on reference:
+        https://cloud.google.com/compute/docs/reference/beta/firewalls and
+        https://cloud.google.com/compute/docs/vpc/firewalls#gcp_firewall_rule_summary_table
+        If add_rule_callback is set, this will also confirm that
+        add_rule_callback returns True for the rule, otherwise it will not add
+        the rule.
+
+        Returns:
+          bool: If rule is valid.
+
+        Raises:
+          DuplicateFirewallRuleNameError: Two or more rules have the same name.
+          InvalidFirewallRuleError: One or more rules failed validation.
+        """
+        self._validate_keys()
+        self._validate_direction()
+        self._validate_priority()
+        if not self.firewall_action:
+            raise InvalidFirewallRuleError('Rule missing action "%s"' % self)
+        else:
+            self.firewall_action.validate()
+
+        # TODO: Verify rule name matches regex of allowed
+        # names from reference
+
+        return True
+
+    def _validate_keys(self):
+        """Checks that required keys and value restrictions.
+
+        Required fields: name and network
+        Length restrictions:
+          * name <= 63 characters
+          * <= 256 values:
+            sourceRanges, sourceTags, targetTags, destinationRanges
+
+        Raises:
+          InvalidFirewallRuleError: If keys don't meet requirements.
+        """
+        if not self.name:
+            raise InvalidFirewallRuleError(
+                'Rule missing required field "%s"' % 'name')
+        if not self.network:
+            raise InvalidFirewallRuleError(
+                'Rule missing required field "%s"' % 'network')
+
+        if len(self.name) > 63:
+            raise InvalidFirewallRuleError(
+                'Rule name exceeds length limit of 63 chars: "%s".' %
+                self.name)
+
```

Comment:
source_service_accounts and target_service_accounts can have at most 1 value. Create a new max_1_value_keys to test them.

Reply:
Done